File: src/docforge/openapi/render.py
from typing import Any

from .iter import Endpoint, Schema
from .refs import ref_link


def jsonpointer_encode(s: str) -> str:
    return s.replace("~", "~0").replace("/", "~1")


def _type_str(schema: dict[str, Any]) -> str:
    t = schema.get("type")
    fmt = schema.get("format")
    if t and fmt:
        return f"{t} ({fmt})"
    return t or "any"


def _schema_summary(schema: dict[str, Any], *, from_kind: str) -> str:
    """Render a one-line summary of a schema (for request/response content lines)."""
    if "$ref" in schema:
        label, href = ref_link(schema["$ref"], from_kind=from_kind)
        return f"[{label}]({href})"
    if schema.get("type") == "array":
        items = schema.get("items") or {}
        return f"array of {_schema_summary(items, from_kind=from_kind)}"
    return f"`{_type_str(schema)}`"


def _render_parameters(params: list[dict[str, Any]]) -> list[str]:
    if not params:
        return []
    lines = [
        "## Parameters",
        "",
        "| Name | In | Type | Required | Description |",
        "|------|----|----|----------|-------------|",
    ]
    for p in params:
        name = p.get("name", "")
        loc = p.get("in", "")
        schema = p.get("schema") or {}
        type_s = _type_str(schema)
        required = "yes" if p.get("required") else "no"
        desc = (p.get("description") or "").replace("|", "\\|").replace("\n", " ")
        lines.append(f"| {name} | {loc} | {type_s} | {required} | {desc} |")
    lines.append("")
    return lines


def _render_request_body(body: dict[str, Any] | None) -> list[str]:
    if not body:
        return []
    lines = ["## Request Body", ""]
    required_marker = " (required)" if body.get("required") else ""
    desc = body.get("description")
    if desc:
        lines.append(desc.strip())
        lines.append("")
    for ctype, media in (body.get("content") or {}).items():
        schema = media.get("schema") or {}
        summary = _schema_summary(schema, from_kind="endpoint")
        lines.append(f"`{ctype}`: {summary}{required_marker}")
    lines.append("")
    return lines


def _render_responses(responses: dict[str, Any]) -> list[str]:
    if not responses:
        return []
    lines = ["## Responses", ""]
    for code, resp in responses.items():
        if not isinstance(resp, dict):
            continue
        desc = (resp.get("description") or "").strip()
        lines.append(f"### {code} {desc}".rstrip())
        lines.append("")
        for ctype, media in (resp.get("content") or {}).items():
            schema = media.get("schema") or {}
            summary = _schema_summary(schema, from_kind="endpoint")
            lines.append(f"`{ctype}`: {summary}")
        lines.append("")
    return lines


def render_endpoint(ep: Endpoint, *, spec_filename: str) -> str:
    pointer = f"#/paths/{jsonpointer_encode(ep.path)}/{ep.method}"
    out: list[str] = [
        f"# {ep.method.upper()} {ep.path}",
        "",
        f"Source: {spec_filename}{pointer}",
        "",
    ]
    if ep.tags:
        out.append(f"**Tags:** {', '.join(ep.tags)}")
        out.append("")
    if ep.description:
        out.append(ep.description.strip())
        out.append("")
    elif ep.summary:
        out.append(ep.summary.strip())
        out.append("")

    out.extend(_render_parameters(list(ep.operation.get("parameters") or [])))
    out.extend(_render_request_body(ep.operation.get("requestBody")))
    out.extend(_render_responses(ep.operation.get("responses") or {}))

    while out and out[-1] == "":
        out.pop()
    return "\n".join(out) + "\n"


import json as _json


def _property_type(schema: dict[str, Any]) -> str:
    if "$ref" in schema:
        label, href = ref_link(schema["$ref"], from_kind="schema")
        return f"[{label}]({href})"
    if schema.get("type") == "array":
        items = schema.get("items") or {}
        return f"array of {_property_type(items)}"
    return _type_str(schema)


def render_schema(schema: Schema, *, spec_filename: str) -> str:
    body = schema.body
    out: list[str] = [
        f"# {schema.name}",
        "",
        f"Source: {spec_filename}#/components/schemas/{schema.name}",
        "",
    ]
    desc = (body.get("description") or "").strip()
    if desc:
        out.append(desc)
        out.append("")

    properties = body.get("properties") if body.get("type") == "object" else None
    if properties:
        required = set(body.get("required") or [])
        out.extend([
            "## Properties",
            "",
            "| Name | Type | Required | Description |",
            "|------|------|----------|-------------|",
        ])
        for prop_name, prop in properties.items():
            type_s = _property_type(prop or {})
            req = "yes" if prop_name in required else "no"
            pdesc = ((prop or {}).get("description") or "").replace("|", "\\|").replace("\n", " ")
            out.append(f"| {prop_name} | {type_s} | {req} | {pdesc} |")
        out.append("")
    else:
        out.append("## Definition")
        out.append("")
        out.append("```json")
        out.append(_json.dumps(body, indent=2, ensure_ascii=False))
        out.append("```")
        out.append("")

    while out and out[-1] == "":
        out.pop()
    return "\n".join(out) + "\n"

File: src/openapi/render.ts
import type { Endpoint, Schema } from "./iter.js";
import { refLink, type FromKind } from "./refs.js";

export function jsonpointerEncode(s: string): string {
  return s.replace(/~/g, "~0").replace(/\//g, "~1");
}

function isPlainObject(x: unknown): x is Record<string, unknown> {
  return x !== null && typeof x === "object" && !Array.isArray(x);
}

function typeStr(schema: Record<string, unknown>): string {
  const t = schema.type;
  const fmt = schema.format;
  if (typeof t === "string" && typeof fmt === "string") return `${t} (${fmt})`;
  if (typeof t === "string") return t;
  return "any";
}

function schemaSummary(
  schema: Record<string, unknown>,
  opts: { fromKind: FromKind },
): string {
  if (typeof schema.$ref === "string") {
    const [label, href] = refLink(schema.$ref, opts);
    return `[${label}](${href})`;
  }
  if (schema.type === "array") {
    const items = isPlainObject(schema.items) ? schema.items : {};
    return `array of ${schemaSummary(items, opts)}`;
  }
  return `\`${typeStr(schema)}\``;
}

function renderParameters(params: unknown[]): string[] {
  if (params.length === 0) return [];
  const lines = [
    "## Parameters",
    "",
    "| Name | In | Type | Required | Description |",
    "|------|----|----|----------|-------------|",
  ];
  for (const p of params) {
    if (!isPlainObject(p)) continue;
    const name = typeof p.name === "string" ? p.name : "";
    const loc = typeof p.in === "string" ? p.in : "";
    const schema = isPlainObject(p.schema) ? p.schema : {};
    const tStr = typeStr(schema);
    const required = p.required ? "yes" : "no";
    const descRaw = typeof p.description === "string" ? p.description : "";
    const desc = descRaw.replace(/\|/g, "\\|").replace(/\n/g, " ");
    lines.push(`| ${name} | ${loc} | ${tStr} | ${required} | ${desc} |`);
  }
  lines.push("");
  return lines;
}

function renderRequestBody(body: unknown): string[] {
  if (!isPlainObject(body)) return [];
  const lines = ["## Request Body", ""];
  const requiredMarker = body.required ? " (required)" : "";
  if (typeof body.description === "string" && body.description.trim()) {
    lines.push(body.description.trim());
    lines.push("");
  }
  const content = isPlainObject(body.content) ? body.content : {};
  for (const [ctype, media] of Object.entries(content)) {
    if (!isPlainObject(media)) continue;
    const schema = isPlainObject(media.schema) ? media.schema : {};
    const summary = schemaSummary(schema, { fromKind: "endpoint" });
    lines.push(`\`${ctype}\`: ${summary}${requiredMarker}`);
  }
  lines.push("");
  return lines;
}

function renderResponses(responses: Record<string, unknown>): string[] {
  if (Object.keys(responses).length === 0) return [];
  const lines = ["## Responses", ""];
  for (const [code, resp] of Object.entries(responses)) {
    if (!isPlainObject(resp)) continue;
    const desc = typeof resp.description === "string" ? resp.description.trim() : "";
    lines.push(`### ${code} ${desc}`.trimEnd());
    lines.push("");
    const content = isPlainObject(resp.content) ? resp.content : {};
    for (const [ctype, media] of Object.entries(content)) {
      if (!isPlainObject(media)) continue;
      const schema = isPlainObject(media.schema) ? media.schema : {};
      const summary = schemaSummary(schema, { fromKind: "endpoint" });
      lines.push(`\`${ctype}\`: ${summary}`);
    }
    lines.push("");
  }
  return lines;
}

export function renderEndpoint(
  ep: Endpoint,
  opts: { specFilename: string },
): string {
  const pointer = `#/paths/${jsonpointerEncode(ep.path)}/${ep.method}`;
  const out: string[] = [
    `# ${ep.method.toUpperCase()} ${ep.path}`,
    "",
    `Source: ${opts.specFilename}${pointer}`,
    "",
  ];
  if (ep.tags.length > 0) {
    out.push(`**Tags:** ${ep.tags.join(", ")}`);
    out.push("");
  }
  if (ep.description.trim()) {
    out.push(ep.description.trim());
    out.push("");
  } else if (ep.summary.trim()) {
    out.push(ep.summary.trim());
    out.push("");
  }

  const params = Array.isArray(ep.operation.parameters) ? ep.operation.parameters : [];
  out.push(...renderParameters(params));
  out.push(...renderRequestBody(ep.operation.requestBody));
  const responses = isPlainObject(ep.operation.responses) ? ep.operation.responses : {};
  out.push(...renderResponses(responses));

  while (out.length > 0 && out[out.length - 1] === "") out.pop();
  return out.join("\n") + "\n";
}

function propertyType(schema: Record<string, unknown>): string {
  if (typeof schema.$ref === "string") {
    const [label, href] = refLink(schema.$ref, { fromKind: "schema" });
    return `[${label}](${href})`;
  }
  if (schema.type === "array") {
    const items = isPlainObject(schema.items) ? schema.items : {};
    return `array of ${propertyType(items)}`;
  }
  return typeStr(schema);
}

export function renderSchema(
  sc: Schema,
  opts: { specFilename: string },
): string {
  const body = sc.body;
  const out: string[] = [
    `# ${sc.name}`,
    "",
    `Source: ${opts.specFilename}#/components/schemas/${sc.name}`,
    "",
  ];
  const desc = typeof body.description === "string" ? body.description.trim() : "";
  if (desc) {
    out.push(desc);
    out.push("");
  }

  const properties =
    body.type === "object" && isPlainObject(body.properties)
      ? body.properties
      : null;

  if (properties) {
    const required = new Set(
      Array.isArray(body.required) ? body.required.map(String) : [],
    );
    out.push(
      "## Properties",
      "",
      "| Name | Type | Required | Description |",
      "|------|------|----------|-------------|",
    );
    for (const [propName, prop] of Object.entries(properties)) {
      const p = isPlainObject(prop) ? prop : {};
      const tStr = propertyType(p);
      const req = required.has(propName) ? "yes" : "no";
      const pdescRaw = typeof p.description === "string" ? p.description : "";
      const pdesc = pdescRaw.replace(/\|/g, "\\|").replace(/\n/g, " ");
      out.push(`| ${propName} | ${tStr} | ${req} | ${pdesc} |`);
    }
    out.push("");
  } else {
    out.push("## Definition");
    out.push("");
    out.push("```json");
    out.push(JSON.stringify(body, null, 2));
    out.push("```");
    out.push("");
  }

  while (out.length > 0 && out[out.length - 1] === "") out.pop();
  return out.join("\n") + "\n";
}
